### PR TITLE
Let Ray serialize xarray objects through pickle.

### LINKED ISSUE #748: Ray fails to serialize xarray object.
The following fails (Python 3.6, Anaconda 4.3, MacOS). This requires `pip install xarray`.

```python
import numpy as np
import xarray as xr

import ray

ray.init()

data = xr.DataArray(np.random.randn(2, 3), coords={'x': ['a', 'b']}, dims=('x', 'y'))
ray.put(data)
```

The error is the following.

```python
---------------------------------------------------------------------------
RaySerializationException                 Traceback (most recent call last)
~/Workspace/ray/python/ray/worker.py in store_and_register(self, object_id, value, depth)
    303                 ray.numbuf.store_list(object_id.id(), self.plasma_client.conn,
--> 304                                       [value])
    305                 break

~/Workspace/ray/python/ray/serialization.py in serialize(obj)
    127                                         .format(type(obj)),
--> 128                                         obj)
    129     class_id = type_to_class_id[type(obj)]

RaySerializationException: Ray does not know how to serialize objects of type <class 'xarray.core.dataarray.DataArray'>.

During handling of the above exception, another exception occurred:

RecursionError                            Traceback (most recent call last)
<ipython-input-1-002206058974> in <module>()
      7 
      8 data = xr.DataArray(np.random.randn(2, 3), coords={'x': ['a', 'b']}, dims=('x', 'y'))
----> 9 ray.put(data)

~/Workspace/ray/python/ray/worker.py in put(value, worker)
   1687         object_id = worker.local_scheduler_client.compute_put_id(
   1688             worker.current_task_id, worker.put_index)
-> 1689         worker.put_object(object_id, value)
   1690         worker.put_index += 1
   1691         return object_id

~/Workspace/ray/python/ray/worker.py in put_object(self, object_id, value)
    348         # Serialize and put the object in the object store.
    349         try:
--> 350             self.store_and_register(object_id, value)
    351         except ray.numbuf.numbuf_plasma_object_exists_error as e:
    352             # The object already exists in the object store, so there is no

~/Workspace/ray/python/ray/worker.py in store_and_register(self, object_id, value, depth)
    306             except serialization.RaySerializationException as e:
    307                 try:
--> 308                     _register_class(type(e.example_object))
    309                     warning_message = ("WARNING: Serializing objects of type "
    310                                        "{} by expanding them as dictionaries "

~/Workspace/ray/python/ray/worker.py in _register_class(cls, pickle, worker)
   1544     if not pickle:
   1545         # Raise an exception if cls cannot be serialized efficiently by Ray.
-> 1546         serialization.check_serializable(cls)
   1547         worker.run_function_on_all_workers(register_class_for_serialization)
   1548     else:

~/Workspace/ray/python/ray/serialization.py in check_serializable(cls)
     60                                            "Ray cannot serialize it "
     61                                            "efficiently.".format(cls))
---> 62     if hasattr(obj, "__slots__"):
     63         raise RayNotDictionarySerializable("The class {} uses '__slots__', so "
     64                                            "Ray may not be able to serialize "

~/anaconda3/lib/python3.6/site-packages/xarray/core/common.py in __getattr__(self, name)
    162             # this avoids an infinite loop when pickle looks for the
    163             # __setstate__ attribute before the xarray object is initialized
--> 164             for source in self._attr_sources:
    165                 with suppress(KeyError):
    166                     return source[name]

~/anaconda3/lib/python3.6/site-packages/xarray/core/dataarray.py in _attr_sources(self)
    485     def _attr_sources(self):
    486         """List of places to look-up items for attribute-style access"""
--> 487         return [self.coords, LevelCoordinatesSource(self), self.attrs]
    488 
    489     def __contains__(self, key):

~/anaconda3/lib/python3.6/site-packages/xarray/core/dataarray.py in attrs(self)
    499     def attrs(self):
    500         """Dictionary storing arbitrary metadata with this array."""
--> 501         return self.variable.attrs
    502 
    503     @attrs.setter

~/anaconda3/lib/python3.6/site-packages/xarray/core/dataarray.py in variable(self)
    364     @property
    365     def variable(self):
--> 366         return self._variable
    367 
    368     @property

... last 4 frames repeated, from the frame below ...

~/anaconda3/lib/python3.6/site-packages/xarray/core/common.py in __getattr__(self, name)
    162             # this avoids an infinite loop when pickle looks for the
    163             # __setstate__ attribute before the xarray object is initialized
--> 164             for source in self._attr_sources:
    165                 with suppress(KeyError):
    166                     return source[name]

RecursionError: maximum recursion depth exceeded in comparison
```

### PR DESCRIPTION
**Pull request by ghost**
This PR fixes https://github.com/ray-project/ray/issues/748.

Reference: http://xarray.pydata.org/en/stable/io.html.

### CODE REVIEW

**Comment by pcmoritz** [comment]
ok to test

**Comment by pcmoritz** [comment]
Thanks for the contribution! Can you have a look how this compares performance wise against from_dict and to_dict from http://xarray.pydata.org/en/stable/io.html or potentially even netCDF?

I have a feeling that from_dict and to_dict might be faster at least for deserialization because it can use our efficient arrow serializer for numerical data.

**Comment by uronce-cc** [comment]
ok to test

**Comment by uronce-cc** [comment]
netCDF is used to serialize xarray to or deserialize it from a disk file.

According to the reference, the pickled object's format may not be backward compatible, so for compatibility, the to_dict and from_dict are preferred.

For performance, I did a quick experimentation:

```
In [9]: %timeit -n 10 ray.put(data)
219 ms ± 13.9 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [10]: %timeit -n 10 ray.put(data)
301 ms ± 66.9 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

The above is the to_dict and from_dict version, while the latter is the pickle version.

So I've transformed the PR to use the to_dict and from_dict methods.

**Comment by pcmoritz** [comment]
Thanks, this is the right solution I think! Can you modify the code so it still works if xarray is not installed (see the test failures on travis) by putting a try .. catch around the module import?

File: python/ray/worker.py [hunk 1]
```diff
@@ -18,6 +18,7 @@
 import threading
 import time
 import traceback
+import xarray as xr
```
Comment (robertnishihara) [inline python/ray/worker.py]: Like Philipp said, we don't want Ray to depend on xarray, so let's remove this import and put it somewhere else.

File: python/ray/worker.py [hunk 2]
```diff
@@ -724,6 +725,30 @@ def array_custom_deserializer(serialized_obj):
         custom_serializer=array_custom_serializer,
         custom_deserializer=array_custom_deserializer)
 
+    # Define a custom serializer and deserializer for handling xarray DataArray.
```
Comment (robertnishihara) [inline python/ray/worker.py]: Let's actually do all of these in a big try/except block and also do the `import xarray` here in that block.

**Review by robertnishihara — COMMENTED** [review]
Thanks for the contribution!

The linter complained about the following (at the bottom of https://travis-ci.org/ray-project/ray/jobs/256789979).
```
flake8 --exclude=python/ray/core/src/common/flatbuffers_ep-prefix/,python/ray/core/generated/,src/numbuf/thirdparty/,src/common/format/,doc/source/conf.py
./python/ray/worker.py:728:80: E501 line too long (80 > 79 characters)
./python/ray/worker.py:749:27: F821 undefined name 'xarray_dataset_custom_serializer'
```

**Comment by robertnishihara** [comment]
To elaborate on Philipp's comment, let's add a test to `test/runtest.py`, probably a new one under `SerializationTest` (but it should still be possible to run `runtest.py` even without `xarray` installed). Then we should make sure to pip install xarray in `.travis/install-dependencies.sh`.

**Comment by mitar** [comment]
Hm, but isn't this pretty inefficient? Shouldn't it be better if xarray would go through arrow?

**Pull request closed without merging**